### PR TITLE
Removed WITH_LLVM and replaced protobuf_generate_cpp with more modern protobuf_generate

### PR DESCRIPTION
**Pull request by shenhanc78**
Now we have LLVM source as a submodule, we can always build with LLVM source, so removed the case when WITH_LLVM=Off.

Also replaced "protobuf_generate_cpp" with a more modern "protobuf_generate". On my system that is not Ubuntu , protobuf_generate_cpp throws error. 

Reference: [[1]](https://gitlab.kitware.com/cmake/cmake/-/issues/21228) [[2]](https://stackoverflow.com/questions/52533396/cmake-cant-find-protobuf-protobuf-generate-cpp/64037117#64037117)

### CODE REVIEW

**Comment by snehasish** [comment]
Thanks Han, I will try out the patch on my Ubuntu machine later today and get back to you.

**Comment by shenhanc78** [comment]
Combined this PR with Daniel's PR#200 , this builds all bits (including propeller) successfully on both Ubuntu 22.04 (which has a very old libprotobuf-dev) and Arch Linux (which has the most up-to-date libprotobuf).

**Comment by erozenfeld** [comment]
@shenhanc78 @snehasish This removes `create_gcov` build. I thought the agreement was that GCC AutoFDO tools will live here for now until a new home for them could be found?

**Comment by snehasish** [comment]
I haven't reviewed this patch yet. But yes, we should not be removing the create_gcov build. @shenhanc78 can you please take a look?

**Comment by erozenfeld** [comment]
I also noticed that instructions for build tools for GCC were removed in https://github.com/google/autofdo/commit/f6b1eccf290254bbead70df40e33984e727ef4de . Can they be restored?

**Comment by snehasish** [comment]
> I also noticed that instructions for building tools for GCC were removed in [f6b1ecc](https://github.com/google/autofdo/commit/f6b1eccf290254bbead70df40e33984e727ef4de) . Can they be restored?

I dropped the custom build gcc command and added the text `To build with g++-9 installed from the package repository use `-DCMAKE_C_COMPILER=gcc-9 -DCMAKE_CXX_COMPILER=g++9` instead in the cmake command above.` Do you feel the need to call out the custom build?

**Comment by erozenfeld** [comment]
I think it's confusing that section 2 is named `Build autofdo tools using clang-10` yet includes instructions to build with g++. Also, the prerequisites include both clang and g++, yet only one of them is required. Can the section be renamed and the prerequisites called out correctly?

**Comment by snehasish** [comment]
> I think it's confusing that section 2 is named `Build autofdo tools using clang-10` yet includes instructions to build with g++. Also, the prerequisites include both clang and g++, yet only one of them is required. Can the section be renamed and the prerequisites called out correctly?

Sure, I can make that change once this and #200 is merged.

**Comment by shenhanc78** [comment]
@erozenfeld Added back gcc tools build part. And re-ran all the builds.

**Comment by dhoekwater** [comment]
Just reran tests with Ubuntu 20.04. With and without `WITH_LLVM=On`, all tests pass except for `addr2cu_test`. We get the following output:
```
error: failed to decompress '.debug_info', LLVM was not built with LLVM_ENABLE_ZSTD or did not find zstd at build time
```

**Comment by snehasish** [comment]
> Just reran tests with Ubuntu 20.04. With and without `WITH_LLVM=On`, all tests pass except for `addr2cu_test`. We get the following output:
> 
> ```
> error: failed to decompress '.debug_info', LLVM was not built with LLVM_ENABLE_ZSTD or did not find zstd at build time
> ```

Shouldn't FORCE_ON for the LLVM_ENABLE_ZSTD build mean that the cmake step would fail? I was able to trigger this on my local machine.

File: CMakeLists.txt [hunk 1]
```diff
@@ -799,6 +814,7 @@ function (config_with_llvm)
     DEPENDS prepare_cmds)
 endfunction()
 
+execute_perf_protobuf()
 if (DEFINED WITH_LLVM)
```
Comment (snehasish) [inline CMakeLists.txt:818]: I thought the intent was to remove the WITH_LLVM flag and make it unconditional to build all the tools? Did you revert these changes to support the gcov build?

**Review by snehasish — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (shenhanc78) [reply CMakeLists.txt:818]: This flag still needs to be there, because some source files need to be compiled both with LLVM and without LLVM, so it is not easy to merge the build of "WITH_LLVM" and "WITHOUT_LLVM". We still to do two configs to build two sets of binaries, one for gcc (gcov, etc), the other for create_llvm_prof.

**Comment by dhoekwater** [comment]
Adding `LLVM_ENABLE_ZSTD(FORCE_ON)` actually got me a CMake failure instead of a test failure. I was able to solve it by installing `libzstd-dev` instead of `zstd` (following https://stackoverflow.com/q/59093431).

With that dependency sorted out, all tests are succeeding.

**Comment by shenhanc78** [comment]
> Adding `LLVM_ENABLE_ZSTD(FORCE_ON)` actually got me a CMake failure instead of a test failure. I was able to solve it by installing `libzstd-dev` instead of `zstd` (following https://stackoverflow.com/q/59093431).
> 
> With that dependency sorted out, all tests are succeeding.

Thanks, yes, `set (LLVM_ENABLE_ZSTD FORCE_ON)` also works for me now. (Previous submission has a spurious "," after LLVM_ENABLE_ZSTD, fixed in the latest checkin.

File: CMakeLists.txt (hunk 1 above)
Comment (snehasish) [reply CMakeLists.txt:818]: Ok, then lets make the variable setting more explicit. Can you add something which can only be set to GCOV or LLVM (defaults to LLVM) since that's our primary use case? Also add a print statement during the cmake step to indicate to the user which type of profile tools will be built.

File: CMakeLists.txt [hunk 2]
```diff
@@ -675,9 +686,11 @@ function (config_with_llvm)
     quipper_perf
     status_provider
     symbol_map)
-  add_test(NAME llvm_propeller_program_cfg_proto_builder_test COMMAND llvm_propeller_program_cfg_proto_builder_test)
+  add_test(NAME llvm_propeller_program_cfg_proto_builder_test
```
Comment (snehasish) [inline CMakeLists.txt:689]: Can you revert the formatting only changes?

Comment (shenhanc78) [reply CMakeLists.txt]: Done.

File: CMakeLists.txt (hunk 1 above)
Comment (shenhanc78) [reply CMakeLists.txt]: Done.

File: README.md [hunk 3]
```diff
@@ -13,7 +13,10 @@
     $ cd autofdo
     $ mkdir build
     $ cd build 
-    $ cmake -DCMAKE_C_COMPILER=clang-10 -DCMAKE_CXX_COMPILER=clang++-10 -DCMAKE_BUILD_TYPE=Release ../
+	$ # To build LLVM Propeller support
+    $ cmake -DENABLE_TOOL=LLVM -DCMAKE_C_COMPILER=clang-10 -DCMAKE_CXX_COMPILER=clang++-10 -DCMAKE_BUILD_TYPE=Release ../
+	$ # Or to build tools for GCC
+	$ cmake -DENABLE_TOOL=GCOV -DCMAKE_C_COMPILER=clang-10 -DCMAKE_CXX_COMPILER=clang++-10 -DCMAKE_BUILD_TYPE=Release ../
```
Comment (snehasish) [inline README.md:19]: Please make the indentation consistent. Also add a new line between the LLVM and GCC version invocations.

A couple of other fixes for this file:
* Update the target platform to Ubuntu 22.04
* Update the prerequisites to include libzstd-dev instead of zstd
* Change "Build autofdo tools using clang-10" to "Build autofdo tools"
* Change "To build LLVM Propeller support" to "Build LLVM tools for AutoFDO and Propeller"
* Add the contents of @dhoekwater [comment](https://github.com/google/autofdo/pull/200#issuecomment-2226223676) about cmake support as a separate section for 20.04 

Did you check that both cmake invocations work with gcc?

Comment (shenhanc78) [reply README.md]: Done.

Comment (shenhanc78) [reply README.md]: > Did you check that both cmake invocations work with gcc?

Yes, tested building with Ubuntu 22.04 LTS GCC/G++ (11.4.0), both builds (GCOV/LLVM) succeeded with #210 (already merged).

File: README.md [hunk 4]
```diff
@@ -1,19 +1,32 @@
...
 
 ```
     $ git clone --recursive --depth 1 https://github.com/google/autofdo.git
     $ git checkout v0.20.1
     $ cd autofdo
     $ mkdir build
     $ cd build 
-    $ cmake -DCMAKE_C_COMPILER=clang-10 -DCMAKE_CXX_COMPILER=clang++-10 -DCMAKE_BUILD_TYPE=Release ../
+    $ # Build LLVM tools for AUtoFDO and Propeller
+    $ cmake -DENABLE_TOOL=LLVM -DCMAKE_C_COMPILER=clang-10 -DCMAKE_CXX_COMPILER=clang++-10 -DCMAKE_BUILD_TYPE=Release ../
```
Comment (snehasish) [inline README.md:26]: Can you drop the version suffix, i.e `-10` from clang? That makes the instructions generic for 20.04 and 22.04. 

Also update the gcc text below to drop the suffixes.

**Review by snehasish — APPROVED** [review]
LGTM. Don't forget to squash before merging. Thanks!

File: README.md (hunk 4 above)
Comment (shenhanc78) [reply README.md]: Done dropping the version suffix for clang/gcc.
